### PR TITLE
virttest.guest_agent: add guest-file-open/write/read/close function

### PR DESCRIPTION
**Pull request by xiangchunfu**
1.add open/write/read/close function
id:1468931 1468933 1468932 1468524
Signed-off-by: Xiangchun Fu <xfu@redhat.com>

### CODE REVIEW

**Comment by yanan-fu** [comment]
please make all check pass, and add doc str for the function.

**Comment by yanan-fu** [comment]
Please remove the second commit and update the first one.

**Comment by CongLi** [comment]
It's not a good summary of 'qemu_guest_agent: support another patch guest_agent_read_write_file' in the commit log.

**Comment by xiangchunfu** [comment]
@yanan-fu  Thanks yanan,  updated commit comments.
@CongLi   Thanks CongLi,  use "add guest-file-open/write/read/close function" to replace support_guest_agent_read_write_file_patch.

**Comment by CongLi** [comment]
acked-by: Cong Li <coli@redhat.com>

**Comment by yanan-fu** [comment]
@xiangchunfu  it is better to modify the title of this PR too. and i prefer to add a short doc str for every function.

**Comment by xiangchunfu** [comment]
@yanan-fu  Thanks Yanan.  I modified title. and add dsc doc for very new functions

**Comment by xiangchunfu** [comment]
@luckyh  
Why NACK?  Is it for falling check or? I am checking failling check.

**Comment by luckyh** [comment]
@xiangchunfu please check if you used `self.cmd` correctly, and write the docstring in unified format.

https://github.com/avocado-framework/avocado-vt/blob/master/virttest/virt_vm.py#L998

File: virttest/guest_agent.py [hunk 1]
```diff
@@ -725,3 +725,35 @@ def fsthaw(self, check_status=True):
...
             return ret
         return -1
+
+    def openfile(self, args):
+        """
+        args:specify file's path and mode option when open file
+        """
+        cmd = "guest-file-open"
+        if self.check_has_command(cmd):
+            return self.cmd(cmd, args, debug="Ture", success_resp="False")
```
Comment (balamuruhans) [inline virttest/guest_agent.py]: `debug="True"` here and elsewhere

Comment (balamuruhans) [inline virttest/guest_agent.py]: If success_resp="False", then it self.cmd always returns "" and returning it from def openfile() is not useful

File: virttest/guest_agent.py [hunk 2]
```diff
@@ -725,3 +725,35 @@ def fsthaw(self, check_status=True):
                     raise
             return ret
         return -1
+
+    def openfile(self, args):
+        """
+        args:specify file's path and mode option when open file
```
Comment (balamuruhans) [inline virttest/guest_agent.py]: :param args:

File: virttest/guest_agent.py (hunk 1 above)
Comment (xiangchunfu) [inline virttest/guest_agent.py]: @balamuruhans Thanks for your view.  I used success_resp="True" to replace success_resp="False" and debug = "True" replace debug = "Ture".  I am trying to integrate 4 functions into a function.   I am debugging it.

Comment (balamuruhans) [inline virttest/guest_agent.py]: @xiangchunfu sure, I see still the typo is not corrected, it is debug="True"

File: virttest/guest_agent.py [hunk 3]
```diff
@@ -725,3 +725,35 @@ def fsthaw(self, check_status=True):
...
             return ret
         return -1
+
+    def openfile(self, args):
+        """
+        :params args:specify file's path and mode option when open file
+        """
+        cmd = "guest-file-open"
+        if self.check_has_command(cmd):
+            return self.cmd(cmd, args, debug="Ture", success_resp="False")
```
Comment (balamuruhans) [inline virttest/guest_agent.py]: debug="True"

Comment (xiangchunfu) [inline virttest/guest_agent.py]: changed.  use debug="True" to replace "debug="Ture"

**Comment by xiangchunfu** [comment]
@luckyh Thanks,  I changed self.cmd 's default value and  the docstring in unified format for per function.

File: virttest/guest_agent.py [hunk 4]
```diff
@@ -725,3 +725,42 @@ def fsthaw(self, check_status=True):
                     raise
             return ret
         return -1
+
+    def openfile(self, args):
+        """
+        :params args: specify file's 'path' and 'mode' option with args.
+        :path: /tmp/agentfile
+        :mode: "w+",if file doesn't exist,mode w+ will create a new file
+        """
```
Comment (suqinhuang) [inline virttest/guest_agent.py]: docstring error, no params path and mode

File: virttest/guest_agent.py [hunk 5]
```diff
@@ -725,3 +725,42 @@ def fsthaw(self, check_status=True):
...
+        cmd = "guest-file-open"
+        if self.check_has_command(cmd):
+            return self.cmd(cmd, args, debug=True, success_resp=True)
+
+    def closefile(self, args):
+        """
+        :params args: specify 'handle' option with args.
+        :handle: when open file, handle will be generated.
```
Comment (suqinhuang) [inline virttest/guest_agent.py]: the same, docstring error

File: virttest/guest_agent.py (hunk 4 above)
Comment (xiangchunfu) [reply virttest/guest_agent.py]: @suqinhuang Thanks for your review.  I have updated docstring in 4 functions.

File: virttest/guest_agent.py (hunk 5 above)
Comment (xiangchunfu) [reply virttest/guest_agent.py]: @suqinhuang  updated

**Comment by vivianQizhu** [comment]
These four function can be unified to one function like file_operete(self, args, cmd)

File: virttest/guest_agent.py [hunk 6]
```diff
@@ -725,3 +725,42 @@ def fsthaw(self, check_status=True):
...
+
+    def openfile(self, args):
+        """
+        :params args: It includes file's 'path' and 'mode' option.
+         path = /tmp/agentfile and mode = "w+".
+         if file doesn't exist,mode w+ will create a new file
+        """
+        cmd = "guest-file-open"
+        if self.check_has_command(cmd):
+            return self.cmd(cmd, args, debug=True, success_resp=True)
```
Comment (luckyh) [inline virttest/guest_agent.py]: The default value of `debug` and `success_resp` are `True`, so do not need to set them explicitly.
https://github.com/avocado-framework/avocado-vt/blob/master/virttest/guest_agent.py#L402

File: virttest/guest_agent.py [hunk 7]
```diff
@@ -725,3 +725,42 @@ def fsthaw(self, check_status=True):
                     raise
             return ret
         return -1
+
+    def openfile(self, args):
+        """
```
Comment (luckyh) [inline virttest/guest_agent.py]: Missing description and `:return:`.

File: virttest/guest_agent.py (hunk 6 above)
Comment (xiangchunfu) [reply virttest/guest_agent.py]: @luckyh Thanks,  I have updated related code.

File: virttest/guest_agent.py [hunk 8]
```diff
@@ -725,3 +725,54 @@ def fsthaw(self, check_status=True):
...
+        if self.check_has_command(cmd):
+            return self.cmd(cmd, args)
+
+    def closefile(self, args):
+        """
+        Send guest agent command "guest-file-close" to close a file with eixst handle value.
+
+        :params args: It includes 'handle' option.
+        handle value will be generated when open file.
+        :return: true if the response received, False if return NULL
```
Comment (dinghuiqing) [inline virttest/guest_agent.py]: This function has no return.

Comment (xiangchunfu) [reply virttest/guest_agent.py]: @dinghuiqing @vivianQizhu  I reconstructed this function.  please review it again.

Comment (xiangchunfu) [reply virttest/guest_agent.py]: (1/2) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.qemu_guest_agent.virtio_serial.check_guest_file: PASS (418.97 s)
 (2/2) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.qemu_guest_agent.isa_serial.check_guest_file: PASS (421.96 s)

File: virttest/guest_agent.py [hunk 9]
```diff
@@ -725,3 +725,14 @@ def fsthaw(self, check_status=True):
...
             return ret
         return -1
+
+    def file_operate(self, args, cmd):
+        """
+        Send file operations's guest agent command.
+
+        :params args: It includes guest agent command's parameters.
+        :params cmd: It specify guest agent command.
+        :return: true if the response received, False if return NULL
```
Comment (vivianQizhu) [inline virttest/guest_agent.py]: This description does not match your code

Comment (xiangchunfu) [reply virttest/guest_agent.py]: @vivianQizhu  I updated description.  please review it again. Thanks.

**Comment by vivianQizhu** [comment]
Suggest you move this to https://github.com/autotest/tp-qemu/pull/1056 , I suppose that is not that common here.

**Comment by xiagao** [comment]
I think you'd better seperate them here,not use file_operete().
Because seperate them are more clear and the args are not the same for every cmd.

**Comment by xiangchunfu** [comment]
Xiagao will re-submit related patch.  In order to avoid conflicts, I will close this patch.

**Pull request closed without merging**